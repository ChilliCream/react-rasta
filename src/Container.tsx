// These imports are needed for declaration files
import * as _React from 'react';
import * as _StyledComponents from 'styled-components';
// -------------------------------------------------------------------
import { _getBreakpoints } from './BreakpointMap';
import { BreakpointValue, BreakpointValues, PropertyValue } from './BreakpointValue';
import ThemeProperties, { Theme } from './Theme';
import styled from './StyledComponents';
import { _getGutterWidth, _ensureInjectingGlobal, _render } from './Utilities';

export interface ContainerProperties extends ThemeProperties {
  fluid?: BreakpointValue<boolean>;
  width?: BreakpointValues<number>;
}

const defaultWidth: BreakpointValues<number> = {
  'sm': 540,
  'md': 720,
  'lg': 960,
  'xl': 1140
};

const Container = styled.div`
  width: 100%;
  margin-right: auto;
  margin-left: auto;

  ${
    (props: ContainerProperties) => {
      const width = _getGutterWidth(props.theme);

      return `
        padding-right: ${width}px;
        padding-left: ${width}px;
      `;
    }
  }
  ${
    (props: ContainerProperties) => {
      if (props.fluid) {
        return '';
      } else {
        const breakpoints = _getBreakpoints(props!.theme);
        const renderer = {
          width: (value?: PropertyValue) => _renderWidth(value as number)
        };
        const valueMap = {
          width: props!.width || defaultWidth
        };

        return `
          ${_render(valueMap, breakpoints, renderer)}
        `;
      }
    }
  }
`;

export default Container;

function _renderWidth(width?: number): string {
  if (width != null && typeof width === 'number') {
    return `
      max-width: ${width}px;
    `;
  }

  return '';
}

_ensureInjectingGlobal();