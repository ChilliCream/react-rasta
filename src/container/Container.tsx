import "../utils/bootstrap";
import { ClassAttributes, HTMLAttributes } from "react";
import { StyledComponentClass } from "styled-components";
import ContainerProperties from "./ContainerProperties";
import fullscreen from "./fullscreen";
import renderWidth from "./renderWidth";
import getWidth from "./width";
import { PropertyValue } from "../media";
import Row from "../row";
import { gutterWidth, styled, Theme } from "../theme";
import { css, render } from "../utils";

const Container = styled.div`
  width: 100%;
  margin-right: auto;
  margin-left: auto;

  ${(props: ContainerProperties) => {
    const width = gutterWidth(props.theme);

    return `
      padding-right: ${width}px;
      padding-left: ${width}px;
    `;
  }} ${(props: ContainerProperties) => {
    if (props.fullscreen) {
      fullscreen();

      return css`
        height: 100%;

        > ${Row} {
          height: 100%;
        }
      `;
    } else if (props.fluid) {
      return "";
    } else {
      const renderer = {
        width: (value?: PropertyValue) => renderWidth(value as number)
      };
      const valueMap = {
        width: getWidth(props)
      };

      return render(valueMap, renderer, props!.theme);
    }
  }};
`;

export default Container;
