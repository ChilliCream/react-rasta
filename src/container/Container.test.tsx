import {shallow} from "enzyme";
import "jest-styled-components";
import React from "react";
import Container from "./Container";

describe("<Container />", () => {
  it("should match the snapshot (no properties set)", () => {
    // act
    const result = shallow(<Container />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (fluid property set to true)", () => {
    // act
    const result = shallow(<Container fluid />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (fullscreen property set to true)", () => {
    // act
    const result = shallow(<Container fullscreen />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (width property set to { xs: 100, sm: 200, md: 300 })", () => {
    // act
    const result = shallow(<Container width={{xs: 100, sm: 200, md: 300}} />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (ignores width when fluid is set)", () => {
    // act
    const result = shallow(
      <Container fluid width={{xs: 100, sm: 200, md: 300}} />,
    );

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (ignores fluid and width when fullscreen is set)", () => {
    // act
    const result = shallow(
      <Container fluid fullscreen width={{xs: 100, sm: 200, md: 300}} />,
    );

    // assert
    expect(result).toMatchSnapshot();
  });
});
