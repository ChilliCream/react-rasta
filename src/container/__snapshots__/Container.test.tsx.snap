// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<Container /> should match the snapshot (fluid property set to true) 1`
] = `
.c0 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
/>
`;

exports[
  `<Container /> should match the snapshot (fullscreen property set to true) 1`
] = `
.c0 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
  padding-right: 15px;
  padding-left: 15px;
  height: 100%;
}

.c0 > .sc-bwzfXH {
  height: 100%;
}

<div
  className="c0"
/>
`;

exports[
  `<Container /> should match the snapshot (ignores fluid and width when fullscreen is set) 1`
] = `
.c0 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
  padding-right: 15px;
  padding-left: 15px;
  height: 100%;
}

.c0 > .sc-bwzfXH {
  height: 100%;
}

<div
  className="c0"
  width={
    Object {
      "md": 300,
      "sm": 200,
      "xs": 100,
    }
  }
/>
`;

exports[
  `<Container /> should match the snapshot (ignores width when fluid is set) 1`
] = `
.c0 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
  width={
    Object {
      "md": 300,
      "sm": 200,
      "xs": 100,
    }
  }
/>
`;

exports[`<Container /> should match the snapshot (no properties set) 1`] = `
.c0 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
  padding-right: 15px;
  padding-left: 15px;
}

@media (min-width:576px) {
  .c0 {
    max-width: 540px;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 720px;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: 960px;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: 1140px;
  }
}

<div
  className="c0"
/>
`;

exports[
  `<Container /> should match the snapshot (width property set to { xs: 100, sm: 200, md: 300 }) 1`
] = `
.c0 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
  padding-right: 15px;
  padding-left: 15px;
  max-width: 100px;
}

@media (min-width:576px) {
  .c0 {
    max-width: 200px;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 300px;
  }
}

<div
  className="c0"
  width={
    Object {
      "md": 300,
      "sm": 200,
      "xs": 100,
    }
  }
/>
`;
